Source/Global/global.cpp

```diff
@@ -113,7 +116,38 @@ HRESULT http_singleton::cleanup_async(
 ) noexcept
 {
     std::shared_ptr<http_singleton> singleton{};
-    RETURN_IF_FAILED(singleton_access(singleton_access_mode::cleanup, nullptr, singleton));
+    HRESULT hr = singleton_access(singleton_access_mode::cleanup, nullptr, singleton);
+
+    // if the singleton is still in use, or already cleaned up, fail immediately
+    if (FAILED(hr))
+    {
+        intptr_t hrPtrSize = hr;
+        return XAsyncBegin(
+            async,
+            reinterpret_cast<void*>(hrPtrSize),
+            reinterpret_cast<void*>(cleanup_async),
+            __FUNCTION__,
+            [](XAsyncOp op, const XAsyncProviderData* data)
+        {
+            switch (op)
+            {
+            case XAsyncOp::Begin:
+            {
+                intptr_t hrPtrSize = reinterpret_cast<intptr_t>(data->context);
+                return static_cast<HRESULT>(hrPtrSize);
+            }
+            case XAsyncOp::Cleanup:
+            {
+                return S_OK;
+            }
+            default:
+            {
+                assert(false);
+                return S_OK;
+            }
+            }
+        });
+    }
```

Comment:
outside the scope of this change, but seems like there should be an easier way to return a synchronous error to the XAsync callback.  Seems clunky to have to write a provider just to get the error to flow to caller correctly

Reply:
It would be handy, yes